students/BrianB/lesson06/mailroom4.py

```diff
@@ -0,0 +1,245 @@
+
+    :return: a letter for each donor in donor_data
+    """
+    #for donor, donations in donor_data.items():
+    #    total = sum(donations)
+    #    num = len(donations)
+    total = gen_donation_stats(total)
+    letter = gen_letter(donor, sum(donor_data[donor]))
+    filename = donor.replace(" ", "_").replace(".", "") + ".txt"
+    with open(filename, "w") as postal:
+        print("created a thank you letter for",
+              gen_donation_stats(donor), ":", gen_donation_stats(num),
+
+              "donations totalling", "${:,.2f}"
+              .format(total), "!")
+        postal.write(letter)
+    return total
+
+
+def sort_stats(donations):
+    return sum(donations[1])
+
+
+def gen_stats_report(donors, total, num, average):
+    """
+    generates a report of donor stats
+
+    :return: a formatted matrix of donor stats
+    """
+    donor_stats = []
+    #donor_stats = [(donors, total, num, average)]
+    #for donors, donations in donor_data.items():
+    #    total = sum(donations)
+    #    num = len(donations)
+    #    average = total / num
+    donor_stats.append((donors, total, num, average))
+
+    donor_stats.sort(key=itemgetter(1), reverse=True)
+    header = [
+        f"{'Donor':<10}|{'($) Total Donated':>21}|{'Total Donations':>19}"
+        f"|{'($) Average Donation':>22}|", "-" * 76
+    ]
+    for line in donor_stats:
+        header.append("{:<10}|{:>21,.2f}|{:>19}|{:>22,.2f}|".format(*line))
+        # I want to [header.append] with f-string formatting but I'm
+        #   not getting it to work
+        # header.append(
+        #              f"{donors:<10}|{total:>21,.2f}|{num:>19}|"
+        #              f"{average:>22,.2f}|"
+        #              )
+    return donors, total, num, average, "\n".join(header)
+
+
+def print_donor_report():
+    print(gen_stats_report())
+
+
+def quit():
+    print("quitting")
+    sys.exit()
+
+
+def main_menu():
+    """
+    the main menu of the mailroom
+
+    :return: sends the user everywhere to do everything
+    """
+    answer = int(input("\n"
+                       "What would you like to do?\n"
+                       "Pick One:\n"
+                       "1 - Give a donation\n"
+                       "2 - Create a Report\n"
+                       "3 - Send Thank You's to all donors\n"
+                       "4 - Quit\n"
+                       ">>> "
+                       ))
+    print("You replied: ", answer)
+    answer_dict = {
+                   1: add_to_donations,
+                   2: print_donor_report,
+                   3: send_letter_to_all,
+                   4: quit
+                   }
+    ##########################################################
+    # QUESTION FOR RON: Why does parenthesis after .get work?
+    answer_dict.get(answer, "Please input 1, 2, 3 or 4")()
+    return answer, answer_dict
+
+
+if __name__ == "__main__":
+    print("Welcome to the Mailroom!")
+
+    donor_data = donor_database
+
+    # main_menu()
```

Comment:
Try to remove all commented out code in your PRs before submitting, it's a good habit to get into when writing code!